# DSCI 504: Intro to SQL 

##  Assignment 07 – Advanced PostgreSQL Functions 

Jeane Kayiba

10/13/2024

**Part 1:** <span style="color: rgb(45, 59, 69); font-family: LatoWeb, &quot;Lato Extended&quot;, Lato, &quot;Helvetica Neue&quot;, Helvetica, Arial, sans-serif; font-size: 16px; background-color: rgb(255, 255, 255);">&nbsp;Write a query that finds the total number of sales for the year 2015. Your query cannot use any aggregates. All calculations should be done using the data/time functions in this week's learning. Your output should be in two columns-- year_sales and num_of_sales.</span>

In [2]:
WITH sales_2015 AS (
    SELECT
        ord_date
    FROM
        orders
    WHERE
        EXTRACT(YEAR FROM ord_date) = 2015
)
SELECT
    '2015' AS year_sales,
    ROW_NUMBER() OVER (ORDER BY ord_date DESC) AS num_of_sales
FROM
    sales_2015;



(135 row(s) affected)

Total execution time: 00:00:00.023

year_sales num_of_sales 2015 1 2015 2 2015 3 2015 4 2015 5 2015 6 2015 7 2015 8 2015 9 2015 10 2015 11 2015 12 2015 13 2015 14 2015 15 2015 16 2015 17 2015 18 2015 19 2015 20 2015 21 2015 22 2015 23 2015 24 2015 25 2015 26 2015 27 2015 28 2015 29 2015 30 2015 31 2015 32 2015 33 2015 34 2015 35 2015 36 2015 37 2015 38 2015 39 2015 40 2015 41 2015 42 2015 43 2015 44 2015 45 2015 46 2015 47 2015 48 2015 49 2015 50 2015 51 2015 52 2015 53 2015 54 2015 55 2015 56 2015 57 2015 58 2015 59 2015 60 2015 61 2015 62 2015 63 2015 64 2015 65 2015 66 2015 67 2015 68 2015 69 2015 70 2015 71 2015 72 2015 73 2015 74 2015 75 2015 76 2015 77 2015 78 2015 79 2015 80 2015 81 2015 82 2015 83 2015 84 2015 85 2015 86 2015 87 2015 88 2015 89 2015 90 2015 91 2015 92 2015 93 2015 94 2015 95 2015 96 2015 97 2015 98 2015 99 2015 100 2015 101 2015 102 2015 103 2015 104 2015 105 2015 106 2015 107 2015 108 2015 109 2015 110 2015 111 2015 112 2015 113 2015 114 2015 115 2015 116 2015 117 2015 118 2015 119 2015 120 2015 121 2015 122 2015 123 2015 124 2015 125 2015 126 2015 127 2015 128 2015 129 2015 130 2015 131 2015 132 2015 133 2015 134 2015 135

In [3]:
WITH sales_2015 AS (
    SELECT
        ord_date
    FROM
        orders
    WHERE
        EXTRACT(YEAR FROM ord_date) = 2015
)
SELECT
    '2015' AS year_sales,
    COUNT(*) AS num_of_sales
FROM
    sales_2015;


(1 row(s) affected)

Total execution time: 00:00:00.016

year_sales num_of_sales 2015 135

<span style="color: rgb(45, 59, 69); font-family: LatoWeb, &quot;Lato Extended&quot;, Lato, &quot;Helvetica Neue&quot;, Helvetica, Arial, sans-serif; font-size: 16px; background-color: rgb(255, 255, 255);">Find the monthly sales for the same year. use the same output as above.</span>

In [4]:
WITH monthly_sales_2015 AS (
    SELECT
        EXTRACT(MONTH FROM ord_date) AS month,
        EXTRACT(YEAR FROM ord_date) AS year
    FROM
        orders
    WHERE
        EXTRACT(YEAR FROM ord_date) = 2015
)
SELECT
    CONCAT(year, '-', month) AS year_month_sales,
    ROW_NUMBER() OVER (PARTITION BY month ORDER BY month) AS num_of_sales
FROM
    monthly_sales_2015;


(135 row(s) affected)

Total execution time: 00:00:00.024

year_month_sales num_of_sales 2015-1 1 2015-1 2 2015-1 3 2015-1 4 2015-1 5 2015-1 6 2015-1 7 2015-1 8 2015-1 9 2015-1 10 2015-1 11 2015-1 12 2015-2 1 2015-2 2 2015-2 3 2015-2 4 2015-2 5 2015-2 6 2015-2 7 2015-2 8 2015-2 9 2015-2 10 2015-2 11 2015-3 1 2015-3 2 2015-3 3 2015-3 4 2015-3 5 2015-3 6 2015-3 7 2015-3 8 2015-3 9 2015-3 10 2015-3 11 2015-3 12 2015-4 1 2015-4 2 2015-4 3 2015-4 4 2015-4 5 2015-4 6 2015-4 7 2015-4 8 2015-4 9 2015-4 10 2015-4 11 2015-4 12 2015-4 13 2015-4 14 2015-4 15 2015-4 16 2015-4 17 2015-4 18 2015-4 19 2015-4 20 2015-4 21 2015-4 22 2015-4 23 2015-5 1 2015-5 2 2015-5 3 2015-5 4 2015-5 5 2015-5 6 2015-5 7 2015-5 8 2015-5 9 2015-5 10 2015-6 1 2015-6 2 2015-6 3 2015-6 4 2015-6 5 2015-6 6 2015-6 7 2015-6 8 2015-7 1 2015-7 2 2015-7 3 2015-7 4 2015-7 5 2015-7 6 2015-7 7 2015-7 8 2015-7 9 2015-7 10 2015-8 1 2015-8 2 2015-8 3 2015-8 4 2015-8 5 2015-8 6 2015-8 7 2015-8 8 2015-8 9 2015-9 1 2015-9 2 2015-9 3 2015-9 4 2015-9 5 2015-9 6 2015-9 7 2015-9 8 2015-9 9 2015-10 1 2015-10 2 2015-10 3 2015-10 4 2015-10 5 2015-10 6 2015-10 7 2015-10 8 2015-10 9 2015-10 10 2015-10 11 2015-10 12 2015-10 13 2015-10 14 2015-11 1 2015-11 2 2015-11 3 2015-11 4 2015-11 5 2015-11 6 2015-11 7 2015-12 1 2015-12 2 2015-12 3 2015-12 4 2015-12 5 2015-12 6 2015-12 7 2015-12 8 2015-12 9 2015-12 10

<span style="color: rgb(45, 59, 69); font-family: LatoWeb, &quot;Lato Extended&quot;, Lato, &quot;Helvetica Neue&quot;, Helvetica, Arial, sans-serif; font-size: 16px; background-color: rgb(255, 255, 255);">Write a query to find the sales for all years in the data</span>

In [5]:
WITH yearly_sales AS (
    SELECT
        EXTRACT(YEAR FROM ord_date) AS year
    FROM
        orders
)
SELECT
    year,
    ROW_NUMBER() OVER (PARTITION BY year ORDER BY year) AS num_of_sales
FROM
    yearly_sales
ORDER BY
    year;


(2599 row(s) affected)

Total execution time: 00:00:00.040

year num_of_sales 2000 1 2000 2 2000 3 2000 4 2000 5 2000 6 2000 7 2000 8 2000 9 2000 10 2000 11 2000 12 2000 13 2000 14 2000 15 2000 16 2000 17 2000 18 2000 19 2000 20 2000 21 2000 22 2000 23 2000 24 2000 25 2000 26 2000 27 2000 28 2000 29 2000 30 2000 31 2000 32 2000 33 2000 34 2000 35 2000 36 2000 37 2000 38 2000 39 2000 40 2000 41 2000 42 2000 43 2000 44 2000 45 2000 46 2000 47 2000 48 2000 49 2000 50 2000 51 2000 52 2000 53 2000 54 2000 55 2000 56 2000 57 2000 58 2000 59 2000 60 2000 61 2000 62 2000 63 2000 64 2000 65 2000 66 2000 67 2000 68 2000 69 2000 70 2000 71 2000 72 2000 73 2000 74 2000 75 2000 76 2000 77 2000 78 2000 79 2000 80 2000 81 2000 82 2000 83 2000 84 2000 85 2000 86 2000 87 2000 88 2000 89 2000 90 2000 91 2000 92 2000 93 2000 94 2000 95 2000 96 2000 97 2000 98 2000 99 2000 100 2000 101 2000 102 2000 103 2000 104 2000 105 2000 106 2000 107 2000 108 2000 109 2000 110 2000 111 2000 112 2000 113 2000 114 2000 115 2000 116 2000 117 2000 118 2000 119 2000 120 2000 121 2000 122 2000 123 2000 124 2000 125 2000 126 2000 127 2000 128 2000 129 2000 130 2000 131 2000 132 2000 133 2000 134 2000 135 2000 136 2000 137 2000 138 2000 139 2000 140 2000 141 2000 142 2000 143 2000 144 2000 145 2000 146 2000 147 2000 148 2000 149 2000 150 2001 1 2001 2 2001 3 2001 4 2001 5 2001 6 2001 7 2001 8 2001 9 2001 10 2001 11 2001 12 2001 13 2001 14 2001 15 2001 16 2001 17 2001 18 2001 19 2001 20 2001 21 2001 22 2001 23 2001 24 2001 25 2001 26 2001 27 2001 28 2001 29 2001 30 2001 31 2001 32 2001 33 2001 34 2001 35 2001 36 2001 37 2001 38 2001 39 2001 40 2001 41 2001 42 2001 43 2001 44 2001 45 2001 46 2001 47 2001 48 2001 49 2001 50 2001 51 2001 52 2001 53 2001 54 2001 55 2001 56 2001 57 2001 58 2001 59 2001 60 2001 61 2001 62 2001 63 2001 64 2001 65 2001 66 2001 67 2001 68 2001 69 2001 70 2001 71 2001 72 2001 73 2001 74 2001 75 2001 76 2001 77 2001 78 2001 79 2001 80 2001 81 2001 82 2001 83 2001 84 2001 85 2001 86 2001 87 2001 88 2001 89 2001 90 2001 91 2001 92 2001 93 2001 94 2001 95 2001 96 2001 97 2001 98 2001 99 2001 100 2001 101 2001 102 2001 103 2001 104 2001 105 2001 106 2001 107 2001 108 2001 109 2001 110 2001 111 2001 112 2001 113 2001 114 2001 115 2001 116 2001 117 2001 118 2001 119 2001 120 2001 121 2001 122 2001 123 2001 124 2001 125 2001 126 2001 127 2001 128 2001 129 2001 130 2001 131 2001 132 2001 133 2001 134 2001 135 2001 136 2001 137 2002 1 2002 2 2002 3 2002 4 2002 5 2002 6 2002 7 2002 8 2002 9 2002 10 2002 11 2002 12 2002 13 2002 14 2002 15 2002 16 2002 17 2002 18 2002 19 2002 20 2002 21 2002 22 2002 23 2002 24 2002 25 2002 26 2002 27 2002 28 2002 29 2002 30 2002 31 2002 32 2002 33 2002 34 2002 35 2002 36 2002 37 2002 38 2002 39 2002 40 2002 41 2002 42 2002 43 2002 44 2002 45 2002 46 2002 47 2002 48 2002 49 2002 50 2002 51 2002 52 2002 53 2002 54 2002 55 2002 56 2002 57 2002 58 2002 59 2002 60 2002 61 2002 62 2002 63 2002 64 2002 65 2002 66 2002 67 2002 68 2002 69 2002 70 2002 71 2002 72 2002 73 2002 74 2002 75 2002 76 2002 77 2002 78 2002 79 2002 80 2002 81 2002 82 2002 83 2002 84 2002 85 2002 86 2002 87 2002 88 2002 89 2002 90 2002 91 2002 92 2002 93 2002 94 2002 95 2002 96 2002 97 2002 98 2002 99 2002 100 2002 101 2002 102 2002 103 2002 104 2002 105 2002 106 2002 107 2002 108 2002 109 2002 110 2002 111 2002 112 2002 113 2002 114 2002 115 2002 116 2002 117 2002 118 2002 119 2002 120 2002 121 2002 122 2002 123 2003 1 2003 2 2003 3 2003 4 2003 5 2003 6 2003 7 2003 8 2003 9 2003 10 2003 11 2003 12 2003 13 2003 14 2003 15 2003 16 2003 17 2003 18 2003 19 2003 20 2003 21 2003 22 2003 23 2003 24 2003 25 2003 26 2003 27 2003 28 2003 29 2003 30 2003 31 2003 32 2003 33 2003 34 2003 35 2003 36 2003 37 2003 38 2003 39 2003 40 2003 41 2003 42 2003 43 2003 44 2003 45 2003 46 2003 47 2003 48 2003 49 2003 50 2003 51 2003 52 2003 53 2003 54 2003 55 2003 56 2003 57 2003 58 2003 59 2003 60 2003 61 2003 62 2003 63 2003 64 2003 65 2003 66 2003 67 2003 68 2003 69 2003 70 2003 71 2003 72 2003 73 2003 74 2003 75 2003 76 2003 77 2003 78 2

- How many sales were there in 2010?

138 sales were made in 2010

In [6]:
WITH sales_2010 AS (
    SELECT
        ord_date
    FROM
        orders
    WHERE
        EXTRACT(YEAR FROM ord_date) = 2010
)
SELECT
    '2010' AS year_sales,
    COUNT(*) AS num_of_sales
FROM
    sales_2010;


(1 row(s) affected)

Total execution time: 00:00:00.004

year_sales num_of_sales 2010 138

<span style="color: rgb(45, 59, 69); font-family: LatoWeb, &quot;Lato Extended&quot;, Lato, &quot;Helvetica Neue&quot;, Helvetica, Arial, sans-serif; font-size: 16px; background-color: rgb(255, 255, 255);">Write a query to track customer acquisition in 2018. Display the number of customers acquired each month. Your output should be limited to two columns. You can name the columns as you see fit.</span>

In [7]:
WITH customer_acquisition_2018 AS (
    SELECT
        EXTRACT(MONTH FROM cus_join_date) AS month,
        EXTRACT(YEAR FROM cus_join_date) AS year,
        cus_num
    FROM
        customers
    WHERE
        EXTRACT(YEAR FROM cus_join_date) = 2018
)
SELECT
    CONCAT('2018-', month) AS acquisition_month,
    COUNT(cus_num) AS num_of_customers
FROM
    customer_acquisition_2018
GROUP BY
    month
ORDER BY
    month;



(12 row(s) affected)

Total execution time: 00:00:00.016

acquisition_month num_of_customers 2018-1 7 2018-2 10 2018-3 11 2018-4 8 2018-5 13 2018-6 8 2018-7 6 2018-8 14 2018-9 6 2018-10 9 2018-11 17 2018-12 13

<span style="color: rgb(45, 59, 69); font-family: LatoWeb, &quot;Lato Extended&quot;, Lato, &quot;Helvetica Neue&quot;, Helvetica, Arial, sans-serif; font-size: 16px; background-color: rgb(255, 255, 255);">Write the same query to track customer acquisition over the entire dataset. You can customize your output to your liking.</span>

In [8]:
WITH customer_acquisition_all_years AS (
    SELECT
        EXTRACT(YEAR FROM cus_join_date) AS year,
        EXTRACT(MONTH FROM cus_join_date) AS month,
        cus_num
    FROM
        customers
)
SELECT
    CONCAT(year, '-', LPAD(month::TEXT, 2, '0')) AS acquisition_period,  -- Format year-month as YYYY-MM
    COUNT(cus_num) AS num_of_customers
FROM
    customer_acquisition_all_years
GROUP BY
    year, month
ORDER BY
    year, month;


(236 row(s) affected)

Total execution time: 00:00:00.016

acquisition_period num_of_customers 2001-01 10 2001-02 9 2001-03 13 2001-04 12 2001-05 16 2001-06 8 2001-07 13 2001-08 8 2001-09 15 2001-10 11 2001-11 9 2001-12 13 2002-01 8 2002-02 10 2002-03 9 2002-04 13 2002-05 10 2002-06 13 2002-07 8 2002-08 9 2002-09 10 2002-10 10 2002-11 13 2002-12 9 2003-01 15 2003-02 10 2003-03 16 2003-04 13 2003-05 14 2003-06 12 2003-07 12 2003-08 13 2003-09 12 2003-10 11 2003-11 11 2003-12 14 2004-01 19 2004-02 9 2004-03 9 2004-04 6 2004-05 7 2004-06 13 2004-07 10 2004-08 6 2004-09 12 2004-10 8 2004-11 13 2004-12 12 2005-01 14 2005-02 8 2005-03 11 2005-04 9 2005-05 14 2005-06 11 2005-07 12 2005-08 17 2005-09 20 2005-10 12 2005-11 15 2005-12 8 2006-01 16 2006-02 9 2006-03 10 2006-04 10 2006-05 10 2006-06 13 2006-07 6 2006-08 16 2006-09 12 2006-10 11 2006-11 16 2006-12 17 2007-01 21 2007-02 13 2007-03 13 2007-04 11 2007-05 10 2007-06 15 2007-07 9 2007-08 13 2007-09 9 2007-10 13 2007-11 11 2007-12 6 2008-01 12 2008-02 8 2008-03 10 2008-04 14 2008-05 18 2008-06 11 2008-07 11 2008-08 9 2008-09 9 2008-10 6 2008-11 11 2008-12 8 2009-01 7 2009-02 14 2009-03 14 2009-04 9 2009-05 18 2009-06 11 2009-07 12 2009-08 13 2009-09 7 2009-10 10 2009-11 18 2009-12 16 2010-01 12 2010-02 11 2010-03 7 2010-04 6 2010-05 9 2010-06 15 2010-07 14 2010-08 12 2010-09 11 2010-10 11 2010-11 14 2010-12 9 2011-01 14 2011-02 13 2011-03 12 2011-04 11 2011-05 16 2011-06 5 2011-07 10 2011-08 7 2011-09 9 2011-10 19 2011-11 15 2011-12 14 2012-01 14 2012-02 8 2012-03 11 2012-04 9 2012-05 12 2012-06 11 2012-07 15 2012-08 11 2012-09 10 2012-10 11 2012-11 12 2012-12 11 2013-01 9 2013-02 10 2013-03 15 2013-04 16 2013-05 11 2013-06 8 2013-07 10 2013-08 7 2013-09 10 2013-10 16 2013-11 9 2013-12 11 2014-01 10 2014-02 8 2014-03 9 2014-04 13 2014-05 9 2014-06 7 2014-07 11 2014-08 9 2014-09 11 2014-10 10 2014-11 16 2014-12 6 2015-01 8 2015-02 6 2015-03 11 2015-04 7 2015-05 11 2015-06 10 2015-07 11 2015-08 19 2015-09 13 2015-10 8 2015-11 16 2015-12 12 2016-01 15 2016-02 14 2016-03 9 2016-04 11 2016-05 14 2016-06 6 2016-07 12 2016-08 8 2016-09 17 2016-10 6 2016-11 12 2016-12 10 2017-01 9 2017-02 8 2017-03 10 2017-04 11 2017-05 10 2017-06 24 2017-07 18 2017-08 14 2017-09 10 2017-10 16 2017-11 8 2017-12 17 2018-01 7 2018-02 10 2018-03 11 2018-04 8 2018-05 13 2018-06 8 2018-07 6 2018-08 14 2018-09 6 2018-10 9 2018-11 17 2018-12 13 2019-01 10 2019-02 12 2019-03 8 2019-04 14 2019-05 16 2019-06 13 2019-07 8 2019-08 10 2019-09 7 2019-10 11 2019-11 13 2019-12 18 2020-01 1 2020-03 0 2020-04 0 2020-06 0 2020-08 0 2020-09 0 2020-11 0 2020-12 0

<span style="font-size: 16px;">W</span><span style="background-color: rgb(255, 255, 255); color: rgb(45, 59, 69); font-family: LatoWeb, &quot;Lato Extended&quot;, Lato, &quot;Helvetica Neue&quot;, Helvetica, Arial, sans-serif; font-size: 16px;">hich year had the most customers?</span>

<span style="font-size: 16px;">2017&nbsp;</span>

Write a query to find the day of week the customer with id 1800 joined. Output the customer's table ID, customer ID, lastname and day of the week they joined.

In [20]:
SELECT
    cus_id,
    cus_num,
    cus_last_name,
    TO_CHAR(cus_join_date, 'Day') AS day_of_week
FROM
    customers
WHERE
    cus_id = 1800;


(1 row(s) affected)

Total execution time: 00:00:00

cus_id cus_num cus_last_name day_of_week 1800 Dic56521 Dickerson Thursday

W<span style="color: rgb(45, 59, 69); font-family: LatoWeb, &quot;Lato Extended&quot;, Lato, &quot;Helvetica Neue&quot;, Helvetica, Arial, sans-serif; font-size: 16px; background-color: rgb(255, 255, 255);">hat is the last name of customer 1800 and dow of the week customer 1800 joined?</span>

<span style="color: rgb(45, 59, 69); font-family: LatoWeb, &quot;Lato Extended&quot;, Lato, &quot;Helvetica Neue&quot;, Helvetica, Arial, sans-serif; font-size: 16px; background-color: rgb(255, 255, 255);">Dickerson joined Thursday&nbsp;</span>

**Part 2:** Determine the number of days between a customer joining and the delivery of their order. Limit your query to the year 2018.

In [21]:
SELECT
    c.cus_id,
    c.cus_num,
    c.cus_last_name,
    o.ord_ship_date,
    c.cus_join_date,
    o.ord_id,
    (o.ord_ship_date - c.cus_join_date) AS days_between
FROM
    customers c
JOIN
    orders o ON c.cus_id = o.cus_id  -- Assuming a relationship between customer and order
WHERE
    EXTRACT(YEAR FROM o.ord_ship_date) = 2018;


(132 row(s) affected)

Total execution time: 00:00:00.026

cus_id cus_num cus_last_name ord_ship_date cus_join_date ord_id days_between 27 Jon59039 Jones 2018-08-16 2003-01-17 5920416 5690 66 Man33420 Manchester 2018-01-26 2019-09-14 6416425 -596 90 Pre57644 Presten 2018-06-13 2019-08-12 2358239 -425 120 Fre53754 French 2018-06-20 2005-03-04 5037369 4856 126 Wil30699 Williamson 2018-10-26 2004-09-06 6960628 5163 183 Smi84773 Smith 2018-07-28 2018-08-10 1985674 -13 189 Say81419 Saylor 2018-04-12 2013-09-17 3695577 1668 195 Man93269 Manchester 2018-06-13 2003-09-03 4924882 5397 215 Smi99812 Smith 2018-09-20 2013-06-30 6562501 1908 217 Joh87634 Johnasen 2018-02-03 2014-04-04 6755650 1401 18 Bro84611 Brooks 2018-02-26 2007-06-17 1724480 3907 240 Row85849 Rowley 2018-06-23 2004-01-07 4366976 5281 263 San42212 Sanchez 2018-04-15 2007-09-06 3951017 3874 308 Ric4760 Rice 2018-09-02 2007-02-20 2508319 4212 321 Ban14334 Bangeson 2018-09-25 2004-05-08 5622284 5253 344 Coo38004 Cooper 2018-04-26 2011-10-02 2974760 2398 347 Jac8664 Jackson 2018-03-07 2013-10-31 1527511 1588 359 Smi26069 Smith 2018-04-17 2017-06-16 5249948 305 370 Ste269 Stevens 2018-05-02 2014-11-27 2160272 1252 439 Man15009 Manchester 2018-04-22 2014-08-24 2612228 1337 462 Smi46223 Smith 2018-02-09 2009-08-11 2615397 3104 476 Aar14232 Aaronson 2018-02-21 2015-09-11 3190336 894 540 Tho9358 Thomason 2018-04-15 2004-09-25 4352288 4950 547 Row35530 Rowley 2018-04-12 2008-05-30 6842420 3604 566 Fre47659 French 2018-02-24 2002-11-14 3600012 5581 578 Ste31874 Stevens 2018-03-06 2003-03-19 1247344 5466 659 Rid6541 Ridgley 2018-07-13 2019-11-14 5415527 -489 707 She26089 Shelton 2018-07-22 2008-11-06 7223848 3545 758 Man18539 Manchester 2018-07-12 2007-05-21 3335994 4070 783 San17058 Sanchez 2018-04-17 2009-11-04 3938629 3086 788 App91835 Applewood 2018-11-25 2002-02-13 4043884 6129 806 Aar5225 Aaronson 2018-08-02 2010-09-20 6964456 2873 819 Coo88227 Cooper 2018-03-26 2001-06-22 5465904 6121 866 Smi30973 Smith-Warren 2018-07-02 2016-10-28 4136230 612 867 Ban37125 Bangeson 2018-07-23 2005-06-23 4409354 4778 891 Goo34132 Goode 2018-11-18 2016-01-24 3442266 1029 919 Gon20205 Gonzales 2018-04-28 2018-03-14 5396896 45 948 Del34352 Deleon 2018-06-25 2001-11-02 6831036 6079 956 Cal39333 Calahan 2018-08-23 2017-04-07 2097092 503 979 Man16461 Manning 2018-04-15 2006-11-09 1438841 4175 1006 Rey40906 Reynolds 2018-10-17 2013-01-01 1134407 2115 1017 Ric86182 Rice 2018-09-04 2001-05-07 1470833 6329 1028 Mau23204 Maurer 2018-02-05 2009-02-14 4923506 3278 1052 Jim24148 Jiminez 2018-01-12 2009-04-20 5923911 3189 1091 Moo57211 Moore 2018-10-13 2017-06-12 4807740 488 1105 Mau86047 Maurer 2018-01-25 2018-06-14 4811446 -140 1112 San47199 Sanchez 2018-11-15 2007-04-09 4019530 4238 1115 Etz74121 Etzler 2018-03-06 2012-06-14 7150691 2091 1121 Jim75660 Jiminez 2018-02-12 2001-09-06 2528103 6003 1126 Rob12788 Robinson 2018-02-04 2002-06-24 5562839 5704 1134 She99098 Shelton 2018-07-15 2005-09-20 2018377 4681 1140 Pre77308 Presley 2018-06-05 2005-09-08 1797451 4653 1153 Man71965 Manchester 2018-07-21 2013-04-02 5847656 1936 1208 Jon48136 Jones 2018-11-10 2009-11-02 3048373 3295 1217 Pre83638 Presley 2018-03-04 2010-05-01 6582795 2864 1229 Fre39803 French 2018-08-10 2007-09-11 3612146 3986 1230 Goo2339 Goode 2018-02-09 2010-11-16 1532917 2642 1255 Coo931 Cooper 2018-09-11 2016-03-27 7107983 898 1257 Smi94995 Smith 2018-02-21 2014-02-10 5023755 1472 1277 Smi70649 Smith 2018-12-21 2005-11-14 1317867 4785 1293 Con9699 Connery 2018-11-25 2019-04-16 4304070 -142 1297 Jon71012 Jones 2018-08-19 2012-01-07 3823807 2416 1305 Tan55919 Tankersly 2018-01-18 2002-02-28 5798754 5803 1313 Boo3995 Booker 2018-12-07 2003-12-23 3967475 5463 1314 Con94549 Connery 2018-03-14 2008-10-10 2823551 3442 1322 Joh79883 Johnasen 2018-11-29 2012-05-03 1213381 2401 1325 Jac60881 Jackson 2018-05-13 2009-07-14 3648251 3225 1334 Jor22698 Jordan 2018-03-22 2003-09-08 6017416 5309 1348 Moo86303 Moore 2018-05-02 2017-01-17 4375558 470 1349 San87554 Sanchez 2018-05-24 2003-03-16 3807129 5548 1359 Tan43

<span style="color: rgb(45, 59, 69); font-family: LatoWeb, &quot;Lato Extended&quot;, Lato, &quot;Helvetica Neue&quot;, Helvetica, Arial, sans-serif; font-size: 16px; background-color: rgb(255, 255, 255);">what was the shortest and longest delay between acquisition and delivery in 2016?</span>

<span style="color: rgb(45, 59, 69); font-family: LatoWeb, &quot;Lato Extended&quot;, Lato, &quot;Helvetica Neue&quot;, Helvetica, Arial, sans-serif; font-size: 16px; background-color: rgb(255, 255, 255);">Shortes delay was -1298 and the longest delay&nbsp; was 5757</span>

**Part 3:** <span style="color: rgb(45, 59, 69); font-family: LatoWeb, &quot;Lato Extended&quot;, Lato, &quot;Helvetica Neue&quot;, Helvetica, Arial, sans-serif; font-size: 16px; background-color: rgb(255, 255, 255);">&nbsp;Perform a query to find all customers with orders fulfilled through the Dallas warehouse. This should be a simple query with no joins. Order your results by the total value of each customer's order.</span>

In [28]:
SELECT
    cus_id,
    cus_last_name,
    tot_ord_value
FROM
    customers
ORDER BY
    tot_ord_value DESC;






(2619 row(s) affected)

Total execution time: 00:00:00.032

cus_id cus_last_name tot_ord_value 1101 Connery 35905 694 Cooper 35902 1353 Rice 35887 2193 Smith 35881 318 Conley 35879 1868 Roberts 35877 1153 Manchester 35854 1461 Hart 35845 1412 Sherman 35831 1960 Jamieson 35829 961 Jackson 35821 1854 French 35808 2096 Thomason 35794 1625 Tankersly 35782 1453 Moore 35778 1933 Barrister 35762 2366 Smith-Warren 35762 1306 Peters 35741 1020 Jamieson 35713 1894 Manchester 35697 2432 Jiminez 35672 559 Peters 35672 2387 Garth 35669 1586 Reynolds 35666 1095 Slain 35662 2024 Willis 35659 1170 Smith 35648 1401 Goode 35637 893 Rowley 35633 2223 Brooks 35627 877 Banks 35614 41 Robinson 35560 1840 Smith-Warren 35557 1097 Ridgley 35546 2512 MacBain 35534 2162 Williamson 35525 637 Acosta 35519 1760 Willis 35469 1233 Hart 35468 535 Calahan 35457 987 Willis 35456 522 Aaronson 35421 950 Applewood 35413 1701 Ridgley 35391 677 Jones 35361 2483 Dewey 35355 1285 Tankersly 35308 571 Reynolds 35304 1130 Hart 35280 819 Cooper 35272 1123 Connery 35257 1917 Riochardson 35245 202 Shelton 35241 253 Tompsen 35230 1163 Hart 35221 1589 Rowley 35208 394 Maurer 35206 138 Ramirez 35181 1380 Dewey 35162 2079 Applewood 35162 457 Jackson 35132 422 Etzler 35129 2254 Moore 35117 455 Willis 35097 2565 Acosta 35058 2070 Presten 35049 147 Acosta 35048 2422 Dawson 35038 429 French 35034 237 Ridgley 35021 1682 Aaronson 35020 1729 Rice 35011 1018 Calahan 35006 2152 Deleon 34989 217 Johnasen 34976 2090 Peters 34955 65 Smith-Warren 34927 205 Stevens 34913 57 Rowley 34904 2286 Smith 34859 2163 French 34855 1110 Slain 34847 1961 Archibald 34839 386 Calahan 34806 754 Tankersly 34805 2545 Archibald 34801 1075 Smith 34799 1094 Gorman 34778 356 Delacruz 34773 155 Cooper 34755 377 Presley 34742 936 Seiler 34724 1788 Jones 34718 1054 Smith-Warren 34677 1042 Reynolds 34653 746 Ridgley 34645 454 Williamson 34632 1943 Williamson 34629 576 Sherman 34629 1671 Booker 34624 2143 Archibald 34618 157 Manchester 34614 2398 Rice 34612 1517 Riochardson 34597 47 Manchester 34586 757 Williamson 34584 1634 Applewood 34564 1515 Jackson 34561 538 Ramirez 34524 1982 Booker 34515 311 Riochardson 34511 1014 Monk 34510 487 Saylor 34488 1235 Manchester 34483 1695 Johnasen 34454 2248 Harper 34447 560 Delapena 34437 937 Hart 34430 504 Connery 34430 966 Banks 34416 2463 Brooks 34402 2331 Connery 34400 1047 Willis 34399 1704 Peters 34389 148 Applewood 34335 565 Ellis 34325 1214 Sherman 34312 2260 Delapena 34310 1112 Sanchez 34306 397 Conley 34295 1931 Cooper 34286 263 Sanchez 34281 2326 Moore 34275 1786 Goode 34272 419 Garth 34266 1483 Jamieson 34260 1224 Aaronson 34256 2201 Harper 34250 2066 Reynolds 34224 2503 Sherman 34163 72 Rice 34148 2259 French 34140 27 Jones 34110 1126 Robinson 34098 2142 Jiminez 34072 1698 Maurer 34054 2580 Ramirez 34028 1221 Aaronson 34027 2054 Reynolds 34007 79 Roberts 33987 240 Rowley 33953 1949 Cooper 33951 1622 Rowley 33938 1853 Harper 33931 1136 Booker 33927 2384 Cooper 33926 2490 Delapena 33923 1478 French 33913 2411 Presten 33906 1158 Dawson 33893 734 Dawson 33869 2489 French 33840 2576 Delapena 33817 667 Acosta 33791 524 Applewood 33753 2166 Ridgley 33740 514 Jones 33740 1562 Willis 33726 617 Jones 33707 2442 Gorman 33678 696 Bangeson 33655 1871 Ellis 33647 1526 Saylor 33646 766 Calahan 33630 26 Rice 33629 983 Goode 33623 1135 Booker 33616 390 Sherman 33610 1590 Saylor 33592 1532 Delapena 33582 410 Connery 33560 2413 Thomason 33547 231 Harper 33525 2052 Smith-Warren 33511 1001 Thomason 33499 1355 Rice 33487 1697 Archibald 33481 2165 Robinson 33364 2072 Riochardson 33299 954 Goode 33296 2224 Roberts 33291 2500 Moore 33285 825 Jordan 33273 1510 Thomason 33256 1495 Ridgley 33255 1987 Barrister 33247 1116 Gorman 33245 1706 Garcia 33217 1964 Smith 33152 662 Seiler 33131 1665 Jones 33110 408 Manchester 33106 199 Jamieson 33064 1547 Shelton 33047 1985 Jiminez 33038 37 Delacruz 33024 592 Dawson 33018 2081 Garcia 33000 1657 Peters 32995 1685 Aaronson 32968 32 Cooper 32966 1435 Reynolds 32954 2117 Tankersly 32932 2179 Harper 32914 714 Etzler 3

**Part 4**: Shimano has reached out to OPC and has asked that OPC provide a listing of all customers that purchased a bike with a combination of Shimano's race components XTR due to a critical recall issuance. They have concerns that not all purchasers have registered their bikes or that they may be out of warranty. Write a query that outputs the following elements to a table that can be exported and provided to Shimano. The table should be stored in the public schema.

- Customer ID
- Customer Last Name
- Customer First Name
- Customer State
- Customer Phone Number

In [33]:
-- Drop the existing table
DROP TABLE IF EXISTS public.shimano_recall_customers;

-- Recreate the table and insert data
CREATE TABLE public.shimano_recall_customers (
    cus_id INT,
    cus_last_name VARCHAR(255),
    cus_first_name VARCHAR(255),
    cus_state VARCHAR(255),
    cus_phone VARCHAR(255)
);

-- Insert the relevant customer data
INSERT INTO public.shimano_recall_customers (cus_id, cus_last_name, cus_first_name, cus_state, cus_phone)
SELECT
    c.cus_id,
    c.cus_last_name,
    c.cus_first_name,
    c.cus_state,
    c.cus_phone
FROM
    customers c
JOIN
    orders o ON c.cus_id = o.cus_id
JOIN
    products p ON o.prod_id = p.prod_id
WHERE
    p.prod_description ILIKE '%Shimano XTR%';








DROP TABLE

CREATE TABLE

INSERT 0 0

Total execution time: 00:00:00.040

In [56]:
SELECT
    c.cus_id,
    c.cus_last_name,
    c.cus_first_name,
    c.cus_state,
    c.cus_phone
FROM
    customers c
JOIN
    orders o ON c.cus_id = o.cus_id
JOIN
    productbuilds pb ON o.prod_id = pb.prod_id
JOIN
    builds b ON pb.build_id = b.build_id
JOIN
    components comp ON (
        b.comp_brake = CAST(comp.comp_id AS VARCHAR) OR
        b.comp_shifter = CAST(comp.comp_id AS VARCHAR) OR
        b.comp_chain_ring = CAST(comp.comp_id AS VARCHAR) OR
        b.comp_derail = CAST(comp.comp_id AS VARCHAR)
    )
WHERE
    comp.comp_name ILIKE '%Shimano%'


(0 row(s) affected)

Total execution time: 00:00:00.003

cus_id cus_last_name cus_first_name cus_state cus_phone

In [36]:
COPY public.shimano_recall_customers TO 'C:/Users/hp/Downloads/shimano_recall_customers.csv' DELIMITER ',' CSV HEADER;



COPY 0

Total execution time: 00:00:00

In [55]:
SELECT *
FROM components
WHERE comp_name ILIKE '%Shimano%';





        





(25 row(s) affected)

Total execution time: 00:00:00.009

comp_id comp_name comp_cost comp_supplier comp_cat sup_id 46 Shimano Deore M6100 2 114.99 44 Standard Component 44 52 Shimano M6120 140.9 44 Standard Component 44 54 Shimano MT500 47.99 44 Budget Component 44 58 Shimano Deore M6120 4 189 44 Standard Component 44 59 Shimano SLX M7100 187 44 Standard Component 44 61 Shimano XT M8100 169.99 44 Standard Component 44 66 Shimano XTR Trail 275 44 Standard Component 44 67 Shimano XT M8120 181 44 Standard Component 44 206 Shimano XT0M8100 59.99 44 Budget Component 44 210 Shimano XT 64.99 44 Budget Component 44 213 Shimano Deore M6100 32 44 Budget Component 44 214 Shimano XTR 137 44 Standard Component 44 216 Shimano SLX 30 44 Budget Component 44 219 Shimano CS-M8100 190 44 Standard Component 44 224 Shimano SLX CS-M7100 99.6 44 Budget Component 44 225 Shimano XTR M9100 389.99 44 Standard Component 44 226 Shimano Deore M6100 56.99 44 Budget Component 44 227 Shimano SLX 78.99 44 Budget Component 44 228 Shimano XT 114.99 44 Standard Component 44 229 Shimano XTR M9100 180 44 Standard Component 44 63 Shimano Deore M6100 2 172.5 44 Standard Component 44 272 Shimano SLX 34 44 Budget Component 44 273 Shimano XT 45 44 Budget Component 44 274 Shimano XTR 78 44 Budget Component 44 275 Shimano Deore M6100 39 44 Budget Component 44

W<span style="color: rgb(45, 59, 69); font-family: LatoWeb, &quot;Lato Extended&quot;, Lato, &quot;Helvetica Neue&quot;, Helvetica, Arial, sans-serif; font-size: 16px; background-color: rgb(255, 255, 255);">hat is the customer ID of the customer on line 20?</span>

<span style="color: rgb(45, 59, 69); font-family: LatoWeb, &quot;Lato Extended&quot;, Lato, &quot;Helvetica Neue&quot;, Helvetica, Arial, sans-serif; font-size: 16px; background-color: rgb(255, 255, 255);">229</span>

**Part 5:** Write a query that will aggregate the total of all bikes sold. Group the output by the bike manufacturer. Your output should contain the following elements:

- Bike Manufacturer
- Bike Name
- Bike Description
- Bike Country of Origin
- Total of all sales

In [66]:
SELECT 
    p.prod_manufacturer AS "Bike Manufacturer",
    p.prod_name AS "Bike Name",
    p.prod_description AS "Bike Description",
    p.country_origin AS "Bike Country of Origin",
    SUM(o.order_tot) AS "Total of All Sales"
FROM 
    orders o
JOIN 
    products p ON o.prod_id = p.prod_id
GROUP BY 
    p.prod_manufacturer, p.prod_name, p.prod_description, p.country_origin
ORDER BY 
    "Total of All Sales" DESC;






(68 row(s) affected)

Total execution time: 00:00:00.003

Bike Manufacturer Bike Name Bike Description Bike Country of Origin Total of All Sales 65 Rail Full-Suspension Mountain Bike 2,3 348444.86 25 SB160 Full-Suspension Mountain Bike 3,4 324004.05 61 Occam Full-Suspension Mountain Bike 2,3 319280.00 14 Altitude Full-Suspension Mountain Bike 2,3 313788.49 25 SB150 Full-Suspension Mountain Bike 3,4 306540 64 Gnarvana Full-Suspension Mountain Bike 3,4 280900.42 15 Wreckoning Full-Suspension Mountain Bike 2,3,4 278751.80 10 E-Series Full-Suspension Mountain Bike 3,4 269617.46 66 Firebird Full-Suspension Mountain Bike 2,3,4 259038.39 22 Mojo Full-Suspension Mountain Bike 2,3 255649.16 43 Stereo Full-Suspension Mountain Bike 2,3,4 235657.41 62 Meta Am Full-Suspension Mountain Bike 2,3,4 235024.36 68 SuperFoxy Full-Suspension Mountain Bike 3,4 230741.87 69 Sight Full-Suspension Mountain Bike 1,2,3 226544.40 15 Following Full-Suspension Mountain Bike 2,3,4 225652.19 64 Trail Pistol Full-Suspension Mountain Bike 2,3 219938.50 1 X Trance Full-Suspension Mountain Bike 2,3,4 212526.77 63 Megatower Full-Suspension Mountain Bike 2,3,4 211932.95 62 Clash Full-Suspension Mountain Bike 2,3,4 208997.77 8 Hei Hei Full-Suspension Mountain Bike 2,3 208890.22 64 Shredd Dog Full-Suspension Mountain Bike 3,4 208804.85 64 Megatrail Full-Suspension Mountain Bike 2,3 207981.74 67 Onset Full-Suspension Mountain Bike 2,3, 199980.00 15 Insurgent Full-Suspension Mountain Bike 1,2,3 199298.94 25 SB115 Full-Suspension Mountain Bike 1,2,3 197773.66 20 Tracer Full-Suspension Mountain Bike 1,2 196193.77 26 Stumpjumper Full-Suspension Mountain Bike 2,3 194885.09 26 Enduro Full-Suspension Mountain Bike 2,3 193599.34 22 Ripley Full-Suspension Mountain Bike 2,3 192953.52 14 Instinct Full-Suspension Mountain Bike 2,3 192800.82 25 SB140 Full-Suspension Mountain Bike 2,3,4 188543.5 43 AMD Full-Suspension Mountain Bike 2,3 187747.24 67 Haste Full-Suspension Mountain Bike 2,3 180871.86 18 Mega Full-Suspension Mountain Bike 2,3,4 179285.05 17 Jekyll Full-Suspension Mountain Bike 1,2,3 178030.41 63 Hightower Full-Suspension Mountain Bike 2,3,4 177261.72 18 Reactor Full-Suspension Mountain Bike 3,4 177031.42 68 Foxy Full-Suspension Mountain Bike 2,3 175908.75 8 Process 134 Full-Suspension Mountain Bike 1,2,3 174591.74 14 Element Full-Suspension Mountain Bike 1,2 172432.23 69 Optic Full-Suspension Mountain Bike 2,3 169617.95 25 SB130 Full-Suspension Mountain Bike 2,3,4 167994.75 3 Rip9 Full-Suspension Mountain Bike 2,3 161450.3 64 The Smash Full-Suspension Mountain Bike 3,4 158426.09 22 Ripmo Full-Suspension Mountain Bike 2,3,4 155615.77 3 RKT9 Full-Suspension Mountain Bike 1,2,3 155181.6 10 G-Series Full-Suspension Mountain Bike 4 153678.99 17 Habit Full-Suspension Mountain Bike 2,3 146300.44 61 OIZ Full-Suspension Mountain Bike 1 142216.55 70 Troy Full-Suspension Mountain Bike 2 134170.08 3 Jet9 Full-Suspension Mountain Bike 2,3 132057.13 17 Scalpel Full-Suspension Mountain Bike 1,2 131881.05 7 Define Full-Suspension Mountain Bike 2,3 130694.83 20 Primer Full-Suspension Mountain Bike 2,3,4 127086.05 10 T-Series Full-Suspension Mountain Bike 2,3 124997.62 24 Sensor Full-Suspension Mountain Bike 2,3 124729.08 25 SB100 Full-Suspension Mountain Bike 1,2 123426.42 8 Process 153 Full-Suspension Mountain Bike 2,3,4 118982.00 1 Trance Full-Suspension Mountain Bike 2,3,4 116946.38 24 Force Full-Suspension Mountain Bike 2,3,4 116708.40 22 DV9 Hardtail Mountain Bike 1 113877.15 14 Thunderbolt Full-Suspension Mountain Bike 1 103154.90 7 Snabb Full-Suspension Mountain Bike 2,3 101909.64 20 Sniper Full-Suspension Mountain Bike 1,2 100955.76 18 Scout Hardtail Mountain Bike 2,3 99342.09 63 Tallboy Full-Suspension Mountain Bike 2,3,4 92438.37 3 Air9 Hardtail Mountain Bike 2,3 81477.28 63 Chameleon Hardtail Mountain Bike 1,2,3 71412.49

<span style="color: rgb(45, 59, 69); font-family: LatoWeb, &quot;Lato Extended&quot;, Lato, &quot;Helvetica Neue&quot;, Helvetica, Arial, sans-serif; font-size: 16px; background-color: rgb(255, 255, 255);">Sort your output by the sum of the total order, manufacturer, and the bike name</span>

In [59]:
SELECT 
    p.prod_manufacturer AS "Bike Manufacturer",
    p.prod_name AS "Bike Name",
    p.prod_description AS "Bike Description",
    p.country_origin AS "Bike Country of Origin",
    SUM(o.order_tot) AS "Total of All Sales"
FROM 
    orders o
JOIN 
    products p ON o.prod_id = p.prod_id
GROUP BY 
    p.prod_manufacturer, p.prod_name, p.prod_description, p.country_origin
ORDER BY 
    SUM(o.order_tot) DESC, 
    p.prod_manufacturer,   
    p.prod_name;           


(68 row(s) affected)

Total execution time: 00:00:00.004

Bike Manufacturer Bike Name Bike Description Bike Country of Origin Total of All Sales 65 Rail Full-Suspension Mountain Bike 2,3 348444.86 25 SB160 Full-Suspension Mountain Bike 3,4 324004.05 61 Occam Full-Suspension Mountain Bike 2,3 319280.00 14 Altitude Full-Suspension Mountain Bike 2,3 313788.49 25 SB150 Full-Suspension Mountain Bike 3,4 306540 64 Gnarvana Full-Suspension Mountain Bike 3,4 280900.42 15 Wreckoning Full-Suspension Mountain Bike 2,3,4 278751.80 10 E-Series Full-Suspension Mountain Bike 3,4 269617.46 66 Firebird Full-Suspension Mountain Bike 2,3,4 259038.39 22 Mojo Full-Suspension Mountain Bike 2,3 255649.16 43 Stereo Full-Suspension Mountain Bike 2,3,4 235657.41 62 Meta Am Full-Suspension Mountain Bike 2,3,4 235024.36 68 SuperFoxy Full-Suspension Mountain Bike 3,4 230741.87 69 Sight Full-Suspension Mountain Bike 1,2,3 226544.40 15 Following Full-Suspension Mountain Bike 2,3,4 225652.19 64 Trail Pistol Full-Suspension Mountain Bike 2,3 219938.50 1 X Trance Full-Suspension Mountain Bike 2,3,4 212526.77 63 Megatower Full-Suspension Mountain Bike 2,3,4 211932.95 62 Clash Full-Suspension Mountain Bike 2,3,4 208997.77 8 Hei Hei Full-Suspension Mountain Bike 2,3 208890.22 64 Shredd Dog Full-Suspension Mountain Bike 3,4 208804.85 64 Megatrail Full-Suspension Mountain Bike 2,3 207981.74 67 Onset Full-Suspension Mountain Bike 2,3, 199980.00 15 Insurgent Full-Suspension Mountain Bike 1,2,3 199298.94 25 SB115 Full-Suspension Mountain Bike 1,2,3 197773.66 20 Tracer Full-Suspension Mountain Bike 1,2 196193.77 26 Stumpjumper Full-Suspension Mountain Bike 2,3 194885.09 26 Enduro Full-Suspension Mountain Bike 2,3 193599.34 22 Ripley Full-Suspension Mountain Bike 2,3 192953.52 14 Instinct Full-Suspension Mountain Bike 2,3 192800.82 25 SB140 Full-Suspension Mountain Bike 2,3,4 188543.5 43 AMD Full-Suspension Mountain Bike 2,3 187747.24 67 Haste Full-Suspension Mountain Bike 2,3 180871.86 18 Mega Full-Suspension Mountain Bike 2,3,4 179285.05 17 Jekyll Full-Suspension Mountain Bike 1,2,3 178030.41 63 Hightower Full-Suspension Mountain Bike 2,3,4 177261.72 18 Reactor Full-Suspension Mountain Bike 3,4 177031.42 68 Foxy Full-Suspension Mountain Bike 2,3 175908.75 8 Process 134 Full-Suspension Mountain Bike 1,2,3 174591.74 14 Element Full-Suspension Mountain Bike 1,2 172432.23 69 Optic Full-Suspension Mountain Bike 2,3 169617.95 25 SB130 Full-Suspension Mountain Bike 2,3,4 167994.75 3 Rip9 Full-Suspension Mountain Bike 2,3 161450.3 64 The Smash Full-Suspension Mountain Bike 3,4 158426.09 22 Ripmo Full-Suspension Mountain Bike 2,3,4 155615.77 3 RKT9 Full-Suspension Mountain Bike 1,2,3 155181.6 10 G-Series Full-Suspension Mountain Bike 4 153678.99 17 Habit Full-Suspension Mountain Bike 2,3 146300.44 61 OIZ Full-Suspension Mountain Bike 1 142216.55 70 Troy Full-Suspension Mountain Bike 2 134170.08 3 Jet9 Full-Suspension Mountain Bike 2,3 132057.13 17 Scalpel Full-Suspension Mountain Bike 1,2 131881.05 7 Define Full-Suspension Mountain Bike 2,3 130694.83 20 Primer Full-Suspension Mountain Bike 2,3,4 127086.05 10 T-Series Full-Suspension Mountain Bike 2,3 124997.62 24 Sensor Full-Suspension Mountain Bike 2,3 124729.08 25 SB100 Full-Suspension Mountain Bike 1,2 123426.42 8 Process 153 Full-Suspension Mountain Bike 2,3,4 118982.00 1 Trance Full-Suspension Mountain Bike 2,3,4 116946.38 24 Force Full-Suspension Mountain Bike 2,3,4 116708.40 22 DV9 Hardtail Mountain Bike 1 113877.15 14 Thunderbolt Full-Suspension Mountain Bike 1 103154.90 7 Snabb Full-Suspension Mountain Bike 2,3 101909.64 20 Sniper Full-Suspension Mountain Bike 1,2 100955.76 18 Scout Hardtail Mountain Bike 2,3 99342.09 63 Tallboy Full-Suspension Mountain Bike 2,3,4 92438.37 3 Air9 Hardtail Mountain Bike 2,3 81477.28 63 Chameleon Hardtail Mountain Bike 1,2,3 71412.49

Wh<span style="color: rgb(45, 59, 69); font-family: LatoWeb, &quot;Lato Extended&quot;, Lato, &quot;Helvetica Neue&quot;, Helvetica, Arial, sans-serif; font-size: 16px; background-color: rgb(255, 255, 255);">t was the bike with the most sales?</span>

<span style="font-size: 16px;">Rail</span>

<span style="font-size: 16px;"><br></span>

<span style="background-color: rgb(255, 255, 255); color: rgb(45, 59, 69); font-family: LatoWeb, &quot;Lato Extended&quot;, Lato, &quot;Helvetica Neue&quot;, Helvetica, Arial, sans-serif; font-size: 16px;">What was the total of sales for the Nukeproof Scout?</span>

<span style="background-color: rgb(255, 255, 255); color: rgb(45, 59, 69); font-family: LatoWeb, &quot;Lato Extended&quot;, Lato, &quot;Helvetica Neue&quot;, Helvetica, Arial, sans-serif; font-size: 16px;"><br></span>

In [62]:
SELECT 
    p.prod_manufacturer AS "Bike Manufacturer",
    p.prod_name AS "Bike Name",
    p.prod_description AS "Bike Description",
    p.country_origin AS "Bike Country of Origin",
    SUM(o.order_tot) AS "Total Sales for Nukeproof Scout"
FROM 
    orders o
JOIN 
    products p ON o.prod_id = p.prod_id
WHERE 
    p.prod_name = 'Nukeproof Scout'
GROUP BY 
    p.prod_manufacturer, p.prod_name, p.prod_description, p.country_origin;



(0 row(s) affected)

Total execution time: 00:00:00.002

Bike Manufacturer Bike Name Bike Description Bike Country of Origin Total Sales for Nukeproof Scout

<span style="color: rgb(45, 59, 69); font-family: LatoWeb, &quot;Lato Extended&quot;, Lato, &quot;Helvetica Neue&quot;, Helvetica, Arial, sans-serif; font-size: 16px; background-color: rgb(255, 255, 255);">Write a query to find the total value of all bikes returned by customers. Your output should include the basic customer info, amount of return, and date of return. You can coordinate the output as you see fit but be sure to not include any PII data.</span>

In [70]:
SELECT
    r.cus_id AS "Customer ID",
    r.tot_ret_amnt AS "Return Amount",
    r.ord_id AS "Order ID",
    r.prod_id AS "Product ID",
    r.warehouse_id AS "Warehouse ID"
FROM
    racs r
WHERE
    r.tot_ret_amnt > 0 
ORDER BY
    r.tot_ret_amnt DESC;




(0 row(s) affected)

Total execution time: 00:00:00.002

Customer ID Return Amount Order ID Product ID Warehouse ID

How many customers have returns?

0

What is the highest amount returned by a customer?

Null

What is one unique thing about this query's results?

Null

<span style="color: rgb(45, 59, 69); font-family: LatoWeb, &quot;Lato Extended&quot;, Lato, &quot;Helvetica Neue&quot;, Helvetica, Arial, sans-serif; font-size: 16px; background-color: rgb(255, 255, 255);">Write a query to calculate the average bike price, the sum of all bikes sold, and the difference between the product price and sale price of every bike purchase. You can include any other columns you feel necessary.</span>

In [80]:
SELECT
    p.prod_name AS "Bike Name",
    p.prod_manufacturer AS "Manufacturer",
    p.prod_price AS "Product Price",
    o.order_tot AS "Sale Price",
    o.order_tot - p.prod_price AS "Price Difference",
    AVG(p.prod_price) OVER () AS "Average Bike Price",
    SUM(o.order_tot) OVER () AS "Total Sales"
FROM
    orders o
JOIN
    products p ON o.prod_id = p.prod_id
ORDER BY
    "Price Difference" DESC;


(2599 row(s) affected)

Total execution time: 00:00:00.061

Bike Name Manufacturer Product Price Sale Price Price Difference Average Bike Price Total Sales Tallboy 63 2799 4278.93 1479.93 4587.7039746056175452 12504625.88 SB150 25 8000 8580 580 4587.7039746056175452 12504625.88 SB150 25 8000 8580 580 4587.7039746056175452 12504625.88 SB160 25 7699.99 8258.24 558.25 4587.7039746056175452 12504625.88 SB160 25 7699.99 8258.24 558.25 4587.7039746056175452 12504625.88 E-Series 10 7565.66 8114.17 548.51 4587.7039746056175452 12504625.88 E-Series 10 7565.66 8114.17 548.51 4587.7039746056175452 12504625.88 SB160 25 7699.99 8238.99 539.00 4587.7039746056175452 12504625.88 SB160 25 7699.99 8238.99 539.00 4587.7039746056175452 12504625.88 SB160 25 7699.99 8238.99 539.00 4587.7039746056175452 12504625.88 SB160 25 7699.99 8238.99 539.00 4587.7039746056175452 12504625.88 Wreckoning 15 7399 7935.43 536.43 4587.7039746056175452 12504625.88 Rail 65 7399 7935.43 536.43 4587.7039746056175452 12504625.88 Rail 65 7399 7935.43 536.43 4587.7039746056175452 12504625.88 Rail 65 7399 7935.43 536.43 4587.7039746056175452 12504625.88 E-Series 10 7565.66 8095.26 529.60 4587.7039746056175452 12504625.88 E-Series 10 7565.66 8085.8 520.14 4587.7039746056175452 12504625.88 E-Series 10 7565.66 8085.8 520.14 4587.7039746056175452 12504625.88 SB150 25 8000 8520 520 4587.7039746056175452 12504625.88 SB150 25 8000 8520 520 4587.7039746056175452 12504625.88 Rail 65 7399 7916.93 517.93 4587.7039746056175452 12504625.88 Rail 65 7399 7916.93 517.93 4587.7039746056175452 12504625.88 Wreckoning 15 7399 7916.93 517.93 4587.7039746056175452 12504625.88 Rail 65 7399 7916.93 517.93 4587.7039746056175452 12504625.88 Wreckoning 15 7399 7907.68 508.68 4587.7039746056175452 12504625.88 SB160 25 7699.99 8200.49 500.50 4587.7039746056175452 12504625.88 SB160 25 7699.99 8200.49 500.50 4587.7039746056175452 12504625.88 E-Series 10 7565.66 8057.43 491.77 4587.7039746056175452 12504625.88 Wreckoning 15 7399 7879.94 480.94 4587.7039746056175452 12504625.88 Wreckoning 15 7399 7879.94 480.94 4587.7039746056175452 12504625.88 Rail 65 7399 7879.94 480.94 4587.7039746056175452 12504625.88 Rail 65 7399 7879.94 480.94 4587.7039746056175452 12504625.88 Rail 65 7399 7879.94 480.94 4587.7039746056175452 12504625.88 Rail 65 7399 7879.94 480.94 4587.7039746056175452 12504625.88 SB150 25 8000 8480 480 4587.7039746056175452 12504625.88 SB150 25 8000 8480 480 4587.7039746056175452 12504625.88 SB150 25 8000 8480 480 4587.7039746056175452 12504625.88 SB150 25 8000 8480 480 4587.7039746056175452 12504625.88 SB150 25 8000 8480 480 4587.7039746056175452 12504625.88 SB150 25 8000 8480 480 4587.7039746056175452 12504625.88 SB150 25 8000 8480 480 4587.7039746056175452 12504625.88 Wreckoning 15 7399 7868.84 469.84 4587.7039746056175452 12504625.88 Wreckoning 15 7399 7868.84 469.84 4587.7039746056175452 12504625.88 Megatower 63 6479.99 6949.79 469.80 4587.7039746056175452 12504625.88 Gnarvana 64 6395 6858.64 463.64 4587.7039746056175452 12504625.88 Gnarvana 64 6395 6858.64 463.64 4587.7039746056175452 12504625.88 Gnarvana 64 6395 6858.64 463.64 4587.7039746056175452 12504625.88 SB160 25 7699.99 8161.99 462.00 4587.7039746056175452 12504625.88 SB160 25 7699.99 8161.99 462.00 4587.7039746056175452 12504625.88 SB160 25 7699.99 8161.99 462.00 4587.7039746056175452 12504625.88 SB160 25 7699.99 8161.99 462.00 4587.7039746056175452 12504625.88 SB160 25 7699.99 8161.99 462.00 4587.7039746056175452 12504625.88 SB160 25 7699.99 8161.99 462.00 4587.7039746056175452 12504625.88 SB160 25 7699.99 8161.99 462.00 4587.7039746056175452 12504625.88 SB160 25 7699.99 8161.99 462.00 4587.7039746056175452 12504625.88 SB160 25 7699.99 8161.99 462.00 4587.7039746056175452 12504625.88 SB150 25 8000 8460 460 4587.7039746056175452 12504625.88 SB150 25 8000 8460 460 4587.7039746056175452 12504625.88 E-Series 10 7565.66 8019.6 453.94 4587.7039746056175452 12504625.88 E-Series 10 7565.66 8019.6 453.94 4587.7039746056175452 12504625.88 E-Series 10 7565.66 8019.6 453.94 4587.7039746056175452 1250462

What is one anomaly detected in the query results?

One potential anomaly is the bikes show a  price difference (i.e., the sale price is less than the product price).

What was the total sales costs for Rocky Mountain bikes?

$782,176.44.

In [87]:
SELECT
    p.prod_manufacturer AS "Manufacturer ID",
    s.sup_name AS "Manufacturer Name",
    SUM(o.order_tot) AS "Total Sales for Rocky Mountain"
FROM
    orders o
JOIN
    products p ON o.prod_id = p.prod_id
JOIN
    suppliers s ON p.prod_manufacturer = CAST(s.sup_id AS VARCHAR)
WHERE
    s.sup_name = 'Rocky Mountain'
GROUP BY
    p.prod_manufacturer, s.sup_name;






(1 row(s) affected)

Total execution time: 00:00:00.018

Manufacturer ID Manufacturer Name Total Sales for Rocky Mountain 14 Rocky Mountain 782176.44

**Part 6:** <span style="color: rgb(45, 59, 69); font-family: LatoWeb, &quot;Lato Extended&quot;, Lato, &quot;Helvetica Neue&quot;, Helvetica, Arial, sans-serif; font-size: 16px; background-color: rgb(255, 255, 255);">&nbsp;Discuss your confidence in using advanced techniques such as window functions after completing this assignment. What advantages do they provide over aggregate functions?</span>

Window functions provide greater flexibility, ease of use, and the ability to handle more complex analytical tasks compared to traditional aggregate functions. They are especially useful when you want to retain access to detailed data while still performing advanced calculations. Mastering window functions boosts your confidence and ability to perform more intricate data analysis tasks efficiently.